src/main/scala/beam/agentsim/infrastructure/parking/ParkingStallSampling.scala

```diff
@@ -3,15 +3,100 @@ package beam.agentsim.infrastructure.parking
+import org.matsim.api.core.v01.network.Link
+import org.matsim.api.core.v01.{Coord, Id}
+import org.matsim.core.utils.collections.QuadTree
+
+import scala.jdk.CollectionConverters.collectionAsScalaIterableConverter
+import scala.math.pow
 
 /**
   * sampling methods for randomly generating stall locations from aggregate information
   */
-object ParkingStallSampling {
+object ParkingStallSampling extends ExponentialLazyLogging {
 
   val maxOffsetDistance = 600.0 // TODO: Make this a config parameter
 
+  def linkBasedSampling(
+    rand: Random,
+    requestLocation: Location,
+    maybeLinkQuadTree: Option[QuadTree[Link]],
+    distanceFunction: (Coord, Coord) => Double,
+    availabilityRatio: Double,
+    taz: TAZ,
+    inClosestZone: Boolean,
+    maxDist: Double = maxOffsetDistance
+  ): Location = {
+    maybeLinkQuadTree match {
+      case Some(linkQuadTree) =>
+        val allLinks = linkQuadTree.getDisk(requestLocation.getX, requestLocation.getY, maxDist).asScala
+        val totalLength = allLinks.foldRight(0.0)(_.getLength + _)
+        var currentLength = 0.0
+        val filteredLinks = rand.shuffle(allLinks).takeWhile { lnk =>
+          currentLength += lnk.getLength
+          currentLength <= totalLength * availabilityRatio
+        }
+        Some(filteredLinks)
+          .filter(_.nonEmpty)
+          .map(
+            _.map(lnk => getClosestPointAlongLink(lnk, requestLocation, distanceFunction)).minBy(loc =>
+              distanceFunction(loc, requestLocation)
+            )
+          )
+          .getOrElse {
+            logger.warn(s"Could not find a link for parking request at location: $requestLocation")
+            availabilityAwareSampling(rand, requestLocation, taz, availabilityRatio, inClosestZone)
+          }
+      case _ =>
+        availabilityAwareSampling(rand, requestLocation, taz, availabilityRatio, inClosestZone)
+    }
+  }
+
+  private def getClosestPointAlongLink(
+    link: Link,
+    requestLocation: Location,
+    distanceFunction: (Coord, Coord) => Double
+  ): Location = {
+    val (p1, p2) = (link.getToNode.getCoord, link.getFromNode.getCoord)
+    val (closestPointOption, closestDistance) = List(p1, p2).foldLeft(None: Option[Coord], Double.PositiveInfinity) {
+      (accumulator, endPoint) =>
+        val dist = distanceFunction(endPoint, requestLocation)
+        if (dist < accumulator._2) {
+          (Some(endPoint), dist)
+        } else {
+          accumulator
+        }
+    }
+
+    closestPointOption match {
+      case Some(closestPoint) =>
+        if (closestDistance < 100) {
+          closestPoint
+        } else {
+          val (dx, dy) = (p2.getX - p1.getX, p2.getY - p1.getY) // vector between p1 and p2
+          val (diffx, diffy) =
+            (requestLocation.getX - p1.getX, requestLocation.getY - p1.getY) // vector between p1 and p3
+          val c1 = (dx * diffx + dy * diffy) / (pow(dx, 2) + pow(dy, 2)) // projection of w onto v
```

Comment:
Please wrap in parens...I hate relying on order of operations. If somebody adds something without paying attention then it could screw the whole thing up. Also, make sure all values are doubles or else you'll end up rounding inadvertently (unless that is what you want)